src/webgpu/shader/execution/memory_model/atomicity.spec.ts

```diff
@@ -3,11 +3,244 @@ export const description = `Tests for the atomicity of atomic read-modify-write
+      @size(16) pre_stress_iterations: u32;
+      @size(16) pre_stress_pattern: u32;
+      @size(16) permute_first: u32;
+      @size(16) permute_second: u32;
+      @size(16) testing_workgroups: u32;
+      @size(16) mem_stride: u32;
+      @size(16) location_offset: u32;
+    };
+    
+    @group(0) @binding(0) var<storage, read_write> test_locations : AtomicMemory;
+    @group(0) @binding(1) var<storage, read_write> results : ReadResults;
+    @group(0) @binding(2) var<storage, read_write> shuffled_workgroups : Memory;
+    @group(0) @binding(3) var<storage, read_write> barrier : AtomicMemory;
+    @group(0) @binding(4) var<storage, read_write> scratchpad : Memory;
+    @group(0) @binding(5) var<storage, read_write> scratch_locations : Memory;
+    @group(0) @binding(6) var<uniform> stress_params : StressParamsMemory;
+    
+    fn permute_id(id: u32, factor: u32, mask: u32) -> u32 {
+      return (id * factor) % mask;
+    }
+    
+    fn stripe_workgroup(workgroup_id: u32, local_id: u32) -> u32 {
+      return (workgroup_id + 1u + local_id % (stress_params.testing_workgroups - 1u)) % stress_params.testing_workgroups;
+    }
+    
+    fn spin(limit: u32) {
+      var i : u32 = 0u;
+      var bar_val : u32 = atomicAdd(&barrier.value[0], 1u);
+      loop {
+        if (i == 1024u || bar_val >= limit) {
+          break;
+        }
+        bar_val = atomicAdd(&barrier.value[0], 0u);
+        i = i + 1u;
+      }
+    }
+    
+    fn do_stress(iterations: u32, pattern: u32, workgroup_id: u32) {
+      let addr = scratch_locations.value[workgroup_id];
+      switch(pattern) {
+        case 0u: {
+          for(var i: u32 = 0u; i < iterations; i = i + 1u) {
+            scratchpad.value[addr] = i;
+            scratchpad.value[addr] = i + 1u;
+          }
+        }
+        case 1u: {
+          for(var i: u32 = 0u; i < iterations; i = i + 1u) {
+            scratchpad.value[addr] = i;
+            let tmp1: u32 = scratchpad.value[addr];
+            if (tmp1 > 100000u) {
+              scratchpad.value[addr] = i;
+              break;
+            }
+          }
+        }
+        case 2u: {
+          for(var i: u32 = 0u; i < iterations; i = i + 1u) {
+            let tmp1: u32 = scratchpad.value[addr];
+            if (tmp1 > 100000u) {
+              scratchpad.value[addr] = i;
+              break;
+            }
+            scratchpad.value[addr] = i;
+          }
+        }
+        case 3u: {
+          for(var i: u32 = 0u; i < iterations; i = i + 1u) {
+            let tmp1: u32 = scratchpad.value[addr];
+            if (tmp1 > 100000u) {
+              scratchpad.value[addr] = i;
+              break;
+            }
+            let tmp2: u32 = scratchpad.value[addr];
+            if (tmp2 > 100000u) {
+              scratchpad.value[addr] = i;
+              break;
+            }
+          }
+        }
+        default: {
+          break;
+        }
+      }
+    }
+    
+    let workgroupXSize = 256;
```

Comment:
This 256 is a bit mysterious. How does it relate to the test params?


I guess this "main" has the actual guts of the tests.   Please explain what is being tested here. What do you mean by "atomicity".   Similarly, the other files are testing different kinds of memory behaviours. They need to be explained.

Reply:
Ideally we'd like to define this outside the shader, but when I was writing these tests I couldn't use a pipeline overridable constant for this. Maybe it's ready now?